Rewrite baggage API description to be consistent with trace API (#1168)
open-telemetry/opentelemetry-specification: specification/baggage/api.md

</summary>

- [Overview](#overview)
- [Baggage](#baggage)
  - [Get Value](#get-value)
  - [Get All Values](#get-all-values)
  - [Set Value](#set-value)
  - [Remove Value](#remove-value)
  - [Context Interaction](#context-interaction)
    - [Clear Baggage in the Context](#clear-baggage-in-the-context)
  - [Propagation](#propagation)

>   - [Conflict Resolution](#conflict-resolution)

Reviewer: This TOC is odd - why is Baggage a section next to Overview when the whole file is about Baggage? I think this would make more sense:
```
- [Overview](#overview)
- [Operations](#operations)
  - [Get Value](#get-value)
  - [Get All Values](#get-all-values)
  - [Set Value](#set-value)
  - [Remove Value](#remove-value)
- [Context Interaction](#context-interaction)
  - [Clear Baggage in the Context](#clear-baggage-in-the-context)
- [Propagation](#propagation)
- [Conflict Resolution](#conflict-resolution)
```

- the `Baggage`
- functions to interact with the `Baggage` in a `Context`

The functions described here are one way to approach interacting with the
`Baggage` via having struct/object that represents the entire Baggage content.
Depending on language idioms, a language API MAY implement these functions by
providing the defined functionality interacting purely via the Context.

The Baggage API MUST be fully functional in the absence of an installed SDK.
This is required in order to enable transparent cross-process Baggage
propagation. If a Baggage propagator is installed into the API, it will work
with or without an installed SDK.

## Baggage

`Baggage` is used to annotate telemetry, adding context and information to
metrics, traces, and logs. It is a struct/object type defined by a set of
name/value pairs describing user-defined properties. Each name in `Baggage` MUST

> be associated with exactly one value.

Reviewer: it is worth adding that the value MUST NOT be null (otherwise the semantics of the Get function below will be broken).

Reviewer 2: +1. I suggest we do that in a follow-up though.

Reviewer 2: @bogdandrutu ping ;)

Author: Filed https://github.com/open-telemetry/opentelemetry-specification/issues/1243

The `Baggage` struct/object/ MUST be immutable, so that the containing Context
also remains immutable.

### Baggage operations

#### Get Value

To access the value for a name/value pair set by a prior event, the Baggage API

> SHOULD provide a function that takes the name as input, and returns a value.

Reviewer: ```suggestion
MUST provide a function that takes the name as input, and returns a value.
```

Returns the value associated with the given name, or null if the given name is
not present.

REQUIRED parameters:

`Name` the name to return the value for.

#### Get All Values

Returns the name/value pairs in the `Baggage`. The order of name/value pairs

> MUST NOT be significant. Based on the language specification, the returned

Reviewer: > Based on the language specification

"specification" or "specifics"?

Author: Done.

value can be either an immutable collection or an immutable iterator to the

> collection of name/value pairs in the `Baggage`.

Reviewer: Not sure if in some languages it's more idiomatic to return a sequence of keys rather than k/v pairs. I think either of those should be permitted.

Author: I don't think the current language blocks that.

#### Set Value

> To record the value for a name/value pair, the Baggage API SHOULD provide a

Reviewer: ```suggestion
To record the value for a name/value pair, the Baggage API MUST provide a
```

Reviewer: This was resolved without action. How can we have baggage without Set function, as allowed by "SHOULD".

Author: Done.

function which takes a a name, and a value as input. Returns a new `Baggage`
that contains the new value. Depending on language idioms, a language API MAY
implement these functions by using a `Builder` pattern and exposing a way to
construct a `Builder` from a `Baggage`.

REQUIRED parameters:

> 

Reviewer: worth clarifying that the value MUST NOT be null.

Reviewer: or, rather, that `Set(key, null)` must be treated as equivalent to `Remove(key)` (because not all languages are strongly typed to disallow null as value).

Author: Filed https://github.com/open-telemetry/opentelemetry-specification/issues/1243

- the `Baggage`
- functions to interact with the `Baggage` in a `Context`

The functions described here are one way to approach interacting with the
`Baggage` via having struct/object that represents the entire Baggage content.
Depending on language idioms, a language API MAY implement these functions by
providing the defined functionality interacting purely via the `Context`.

The Baggage API MUST be fully functional in the absence of an installed SDK.
This is required in order to enable transparent cross-process Baggage
propagation. If a Baggage propagator is installed into the API, it will work
with or without an installed SDK.

> `Baggage` is used to annotate telemetry, adding context and information to

Reviewer: Can we move this paragraph to be the very first paragraph under `## Overview`, before `The Baggage API consists of:`?

Author: Done.

metrics, traces, and logs. It is a set of name/value pairs describing
user-defined properties. Each name in `Baggage` MUST be associated with
exactly one value.

The `Baggage` container MUST be immutable, so that the containing `Context`
also remains immutable.

## Operations

### Get Value

To access the value for a name/value pair set by a prior event, the Baggage API
MUST provide a function that takes the name as input, and returns a value
associated with the given name, or null if the given name is not present.

REQUIRED parameters:

`Name` the name to return the value for.

### Get All Values

Returns the name/value pairs in the `Baggage`. The order of name/value pairs
MUST NOT be significant. Based on the language specifics, the returned
value can be either an immutable collection or an iterator on the immutable
collection of name/value pairs in the `Baggage`.

### Set Value

To record the value for a name/value pair, the Baggage API MUST provide a

> function which takes a a name, and a value as input. Returns a new `Baggage`

Reviewer: ```suggestion
function which takes a name, and a value as input. Returns a new `Baggage`
```

- the `Baggage`
- functions to interact with the `Baggage` in a `Context`

The functions described here are one way to approach interacting with the
`Baggage` via having struct/object that represents the entire Baggage content.
Depending on language idioms, a language API MAY implement these functions by
providing the defined functionality interacting purely via the Context.

The Baggage API MUST be fully functional in the absence of an installed SDK.
This is required in order to enable transparent cross-process Baggage
propagation. If a Baggage propagator is installed into the API, it will work
with or without an installed SDK.

## Baggage

`Baggage` is used to annotate telemetry, adding context and information to

> metrics, traces, and logs. It is a struct/object type defined by a set of

Reviewer: ```suggestion
metrics, traces, and logs. It is a set of
```

Author: Done.

name/value pairs describing user-defined properties. Each name in `Baggage` MUST

> be associated with exactly one value.

Reviewer: this paragraph should be at the top in the Overview

Author: Done.

> The `Baggage` struct/object/ MUST be immutable, so that the containing Context

Reviewer: ```suggestion
The `Baggage` container MUST be immutable, so that the containing Context
```

Author: Done.

also remains immutable.

### Get Value

To access the value for a name/value pair set by a prior event, the Baggage API

> MUST provide a function that takes the name as input, and returns a value.
> Returns the value associated with the given name, or null if the given name is
> not present.

Reviewer: ```suggestion
MUST provide a function that takes the name as input, and returns a value
associated with the given name, or null if the given name is not present.
```

Author: Done.

REQUIRED parameters:

`Name` the name to return the value for.

### Get All Values

Returns the name/value pairs in the `Baggage`. The order of name/value pairs
MUST NOT be significant. Based on the language specifics, the returned

> value can be either an immutable collection or an immutable iterator to the

Reviewer: ```suggestion
value can be either an immutable collection or an iterator on the immutable
```

Author: Done.

OPTIONAL parameters:

`Metadata` Optional metadata associated with the name-value pair. This should be
an opaque wrapper for a string with no semantic meaning. Left opaque to allow
for future functionality.

#### Remove Name/Value

> To delete a name/value pair, the Baggage API SHOULD provide a function which

Reviewer: ```suggestion
To delete a name/value pair, the Baggage API MUST provide a function which
```

Reviewer: also resolved without action - why are we changing MUST->SHOULD?

Author: Done.

takes a name as input. Returns a new `Baggage` which no longer contains the
selected name. Depending on language idioms, a language API MAY
implement these functions by using a `Builder` pattern and exposing a way to
construct a `Builder` from a `Baggage`.

REQUIRED parameters:

`Name` the name to remove.

### Clear Baggage in the Context

To avoid sending any name/value pairs to an untrusted process, the Baggage API

> SHOULD provide a way to remove all baggage entries from a context.

Reviewer: SHOULD or MUST?

Reviewer 3: +1 for MUST

Author: Done.

OPTIONAL parameters:

`Metadata` Optional metadata associated with the name-value pair. This should be
an opaque wrapper for a string with no semantic meaning. Left opaque to allow

> for future functionality.

Reviewer: I am curious - does any SDK support that? Feel kinda weird having this "we accept AnyType and don't even know what to do with it".

Author: Not trying to change this now. We can address in a separate PR.

See [Propagators Distribution](../context/api-propagators.md#propagators-distribution)
for how propagators are to be distributed.

Note: The W3C baggage specification does not currently assign semantic meaning
to the optional metadata.

On `extract`, the propagator should store all metadata as a single metadata instance per entry.
On `inject`, the propagator should append the metadata per the W3C specification format.

Notes:

> If the propagator is unable to parse the `baggage` header, `extract` MUST return

Reviewer: * why are we referring to the "`baggage` header" here? Header is determined by the propagator
* is this behavior any different from all propagators, namely no-error/no-op behavior? if not, I am not sure it's worth restating.

Author: This PR does not change this. So I would like to keep them here and address this later.
